### PR TITLE
[MRG] ENH: Add orthogonal power envelope correlation

### PR DESCRIPTION
**Pull request by larsoner**
Adds orthogonal envelope correlation code. Still needs:

- [x] add tests for new functions (e.g., random matrix equiv to @SherazKhan's calculations for orth corr?)
- [x] fix memory problems
- [x] check example works on CircleCI
- [x] add ref to Hipp 2012, Khan 2018
- [x] add reduction argument and allow returning single-epoch matrices

Closes #5389.

Next PR:

- [ ] more methods (don't orth?)
- [ ] add `n_jobs`
- [ ] add `logger` output

Also cleans up a flake8 error in `test_source_estimate.py`, a docstring in `source_estimate.py`, and simplifies `plot_label_source_activations.py`. (I changed these while adding a new `mode` to `extract_label_time_course` which I eventually did not need, but they're nice things to have anyway.

### CODE REVIEW

**Comment by larsoner** [comment]
Looks like GitHub API requests are having a bad time currently, so I'll just paste in what it looks like locally so that people can see what it will eventually hopefully look like on CircleCI:

envcorr

File: mne/connectivity/envelope.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+#          Denis Engemann <denis.engemann@gmail.com>
+#
+# License: BSD (3-clause)
+
+import numpy as np
+
+from ..filter import next_fast_len
+
+
+def orthogonal_envelope_correlation(data):
```
Comment (dengemann) [inline mne/connectivity/envelope.py]: Can we call it `envelope_correlation` or so and make the orthoganalization an optional argument?
I'd then make a private orthogonalization function that we can enrich with more option, e.g., Housholder or QR decomposion for multivariate orthogonalization.

Comment (larsoner) [reply mne/connectivity/envelope.py]: I'll rename but let's save the refactoring for when we actually need it

File: mne/connectivity/envelope.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            raise ValueError('Each entry in data must be 2D, got shape %s'
+                             % (data_.shape,))
+        this_n_nodes, n_times = data_.shape
+        if n_nodes is None:
+            n_nodes = this_n_nodes
+        if this_n_nodes != n_nodes:
+            raise ValueError('n_nodes mismatch between epochs, got %s and %s'
+                             % (n_nodes, this_n_nodes))
+        n_fft = next_fast_len(n_times)
+        data_ = hilbert(data_, N=n_fft, axis=-1)[..., :n_times]
```
[the remark this answers is not in the document]
Comment (larsoner) [reply mne/connectivity/envelope.py]: Done

**Comment by dengemann** [comment]
@larsoner let's not forget to make the orthogonalization an argument such that we can cover the cases of 1) not orthogonalizing 2) bivariate orthogonalization 3) different multivariate orthogonalization

**Comment by larsoner** [comment]
> let's not forget to make the orthogonalization an argument such that we can cover the cases of 1) not orthogonalizing 2) bivariate orthogonalization 3) different multivariate orthogonalization

Can we start with the basic use case for now (orthogonalization) and add these later? As long as we want the current option to be the default, it's future-compatible.

**Comment by larsoner** [comment]
Okay cropping down to 60 sec isn't even good enough. I'll try another dataset to see if it's any better. Also maybe @SherazKhan was right when he mentioned memory requirements can be a problem :) I'll look into optimizing for memory next.

File: mne/connectivity/envelope.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        # to do raw.apply_hilbert if they want)
+        if data_.dtype in (np.float32, np.float64):
+            n_fft = next_fast_len(n_times)
+            data_ = hilbert(data_, N=n_fft, axis=-1)[..., :n_times]
+        if data_.dtype not in (np.complex64, np.complex128):
+            raise ValueError('data.dtype must be float or complex, got %s'
+                             % (data_.dtype,))
+        data_mag = np.abs(data_)
+        data_orth = np.einsum('it,jt->ijt', data_,
+                              data_.conj() / data_mag).imag
```
Comment (dengemann) [inline mne/connectivity/envelope.py]: This computes the the correlation between all variables.
I remember that computing bivariately the upper or lower triangle only saved time and memory.
To be benchmarked, looking into it.

File: examples/connectivity/plot_mne_inverse_envelope_correlation.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+Compute orthogonal envelope correlations of activity in source space
+using resting state CTF data.
+"""
+# Authors: Eric Larson <larson.eric.d@gmail.com>
+#          Sheraz Khan <sheraz@khansheraz.com>
+#          Denis Engemann <denis.engemann@gmail.com>
+#
+# License: BSD (3-clause)
+
```
Comment (dengemann) [inline examples/connectivity/plot_mne_inverse_envelope_correlation.py]: Getting ValueError on running this example:

"/Users/dengeman/mne_data/MNE-brainstorm-data/bst_resting/MEG/bst_resting/subj002_spontaneous_20111102_01_AUX.ds"

File: examples/connectivity/plot_mne_inverse_envelope_correlation.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+##############################################################################
+# Compute label time series and do envelope correlation
+# -----------------------------------------------------
+
+labels = mne.read_labels_from_annot(subject, 'aparc_sub',
+                                    subjects_dir=subjects_dir)
+stcs = apply_inverse_epochs(epochs, inv, lambda2=1. / 9., pick_ori='normal')
+label_ts = mne.extract_label_time_course(
+    stcs, labels, inv['src'], return_generator=True)
+corr = envelope_correlation(label_ts)
```
Comment (dengemann) [inline examples/connectivity/plot_mne_inverse_envelope_correlation.py]: Can we make a matrix plot here with colorbar as a double check that values are as expected.

File: examples/connectivity/plot_mne_inverse_envelope_correlation.py (hunk 4 above)
Comment (larsoner) [reply examples/connectivity/plot_mne_inverse_envelope_correlation.py]: Your dataset is out of date

File: mne/connectivity/envelope.py (hunk 3 above)
Comment (larsoner) [reply mne/connectivity/envelope.py]: Don't bother, I need to change it to row-wise for memory reasons anyway

File: examples/connectivity/plot_mne_inverse_envelope_correlation.py (hunk 5 above)
Comment (larsoner) [reply examples/connectivity/plot_mne_inverse_envelope_correlation.py]: Sure, though I'm not sure how to interpret it. What should the narrative around the plot be?

File: mne/connectivity/envelope.py (hunk 3 above)
Comment (larsoner) [reply mne/connectivity/envelope.py]: (Also, the matrix is not symmetric, so you can't just compute the upper or lower tri. That's why there are the `x=abs(x)` then `(x+x.T)/2` steps)

**Comment by larsoner** [comment]
https://11155-1301584-gh.circle-artifacts.com/0/dev/auto_examples/connectivity/plot_mne_inverse_envelope_correlation.html

**Comment by larsoner** [comment]
Ready for review/merge from my end @SherazKhan @dengemann feel free to look when you get a chance

**Comment by larsoner** [comment]
@agramfort done after consulting with @SherazKhan

**Comment by larsoner** [comment]
https://11158-1301584-gh.circle-artifacts.com/0/dev/auto_examples/connectivity/plot_mne_inverse_envelope_correlation.html#id4

File: mne/connectivity/tests/test_envelope.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import numpy as np
```
Comment (agramfort) [inline mne/connectivity/tests/test_envelope.py]: missing license + authors

File: mne/connectivity/tests/test_utils.py [hunk 7]
```diff
@@ -21,3 +23,31 @@ def test_indices():
...
+    assert_array_equal(degree(corr), np.zeros(10))
+    # more interesting
+    corr = np.array([[0.5, 0.7, 0.4],
+                     [0.1, 0.3, 0.6],
+                     [0.2, 0.8, 0.9]])
+    deg = degree(corr, 1)
+    assert_array_equal(deg, [2, 2, 2])
+    # wanted values obtained with:
+    # import bct
+    # want = bct.degrees_und(bct.utils.threshold_proportional(corr, 0.25) > 0)
```
Comment (agramfort) [inline mne/connectivity/tests/test_utils.py]: ???

**Comment by agramfort** [comment]
besides LGTM

I would be curious to see what this gives on Cam-CAN and it allows to compute a ICA style analysis to estimate "networks"

**Comment by larsoner** [comment]
> I would be curious to see what this gives on Cam-CAN

Glad you asked! I actually made this example using `bst_resting` from some Cam-CAN data, gist [here](https://gist.github.com/larsoner/2dda2c40ae6404a1ef0d027d09e43605) to produce:

run_envcorr_adapted

Which looks a lot like what @SherazKhan and @dengemann made with [their code](https://github.com/SherazKhan/mne-camcan/blob/master/transit_scripts/run_envcorr.py) (upper right image):

sheraz

The awesome thing is that it was robust to all the changes I listed at the top of my gist in preparation for switching to the `bst_resting` dataset.

>  it allows to compute a ICA style analysis to estimate "networks"

If you mean [this](https://www.pnas.org/content/108/40/16783#sec-4), the order there is band-pass, hilbert, magnitude, downsample, ICA, . Here the order is band-pass, hilbert, orthogonalize, correlate. So it's a bit different, unless you want to swap in some sort of correlation for ICA. But I was just thinking we should try it at some point.

**Comment by dengemann** [comment]
@larsoner I am currently updating the brainstorm data. I will then investigate the zeros issue and memory problems and do some tests / benchmarking.

**Comment by agramfort** [comment]
cool

we can play with ICA later.

>

**Comment by larsoner** [comment]
@dengemann feel free to try the gist I linked above, too, you should see the same plot

File: mne/connectivity/tests/test_utils.py (hunk 7 above)
Comment (dengemann) [reply mne/connectivity/tests/test_utils.py]: Leftover from Python version of brain connectivity toolbox Sheraz and I used before.

Comment (larsoner) [reply mne/connectivity/tests/test_utils.py]: I updated the comment to clarify

**Comment by dengemann** [comment]
I will either merge or get back to you with additional queries.

**Comment by larsoner** [comment]
This pull request **introduces 2 alerts** when merging 59525845ffa65da8b5cc223c809706ba4ba0b758 into ec4844e0c6ba40a39030b06253f1adf7e7c55123 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-fc84f9775f5e3cacd9391f0f3cf70183dcdad77d)

**new alerts:**

* 2 for Encoding error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by dengemann** [comment]
example on my machine

screenshot 2019-02-11 21 58 19

output with new PR code plugged in old pipeline

screenshot 2019-02-11 22 13 23

Tests pass and the mysterious bug where all values were zeros disappeared.
Seems we're good to go!

Just one thing @larsoner. Can we avoid computing the median or average inside the function?
I see many cases where you want to have the power envelope matrices for a given epoch, e.g, to compute stats over epochs / comparing it to empty room connectivity. How would you go about that?

**Comment by larsoner** [comment]
> How would you go about that?

In the future we could add `reduction='median' | None` for example

**Comment by dengemann** [comment]
> In the future we could add `reduction='median' | None` for example

So you would not see a problem in that for a next PR given the currently proposed API. I just want to make sure.

**Comment by larsoner** [comment]
@dengemann feel free to merge if you're happy with `combine`

**Comment by agramfort** [comment]
@dengemann honor is yours to merge this one !

**Comment by dengemann** [comment]
Thank you @larsoner !! As always a pleasure to work with you. Let's keep evolving the power envelopes over a few more PRs together in the next month.

**Pull request merged**